### PR TITLE
change PostEditNonblock to take either message ID or outbox ID CORE-9152

### PR DESCRIPTION
**Pull request by mmaxim**
Patch aims to solve the problem of editing pending messages behaving strangely in current builds (normally generates two sends of the same message, slightly altered). cc @chrisnojima @buoyad 

1.) Add a new parameter to `PostEditNonblock`, called `Target` which can be used to specify either a message or outbox ID. The frontend just aims to send whatever it has for the message being edited.
2.) If we get a message ID, then just operate as normal. If we get an outbox ID, we attempt to update the message in the outbox. If we find the outbox ID in the outbox, we update the contents.
3.) If we can't find the outbox ID in the outbox, we fetch the thread and try to find a message with that outbox ID on it, and post an edit.
4.) If we find the outbox ID, but when we update we don't do anything, then do the fetch thread thing as well.
5.) Add a new method to `Deliverer`, called `UpdateInFlight`, which will update the outbox item as well as cancel any outstanding send (in case we are updating the thing we are sending). Note this phase isn't perfect, since we are trying to solve a race between the UI, service, and server.

### CODE REVIEW

File: shared/reducers/chat2.js [hunk 1]
```diff
@@ -214,6 +214,18 @@ const messageMapReducer = (messageMap, action, pendingOutboxToOrdinal) => {
...
     }
+    case Chat2Gen.pendingMessageWasEdited: {
+      const {conversationIDKey, ordinal, text} = action.payload
+      return messageMap.updateIn(
+        [conversationIDKey, ordinal],
+        message =>
+          !message || message.type !== 'text'
+            ? message
+            : message.withMutations(m => {
+                m.set('text', text)
```
Comment (chrisnojima) [inline shared/reducers/chat2.js]: This can just be message.set('text', text)

File: shared/reducers/chat2.js [hunk 2]
```diff
@@ -820,6 +832,7 @@ const rootReducer = (
     case Chat2Gen.messageDelete:
     case Chat2Gen.messageEdit:
     case Chat2Gen.messageWasEdited:
+    case Chat2Gen.pendingMessageWasEdited:
```
Comment (chrisnojima) [inline shared/reducers/chat2.js]: ??

Comment (mmaxim) [reply shared/reducers/chat2.js]: What's wrong here?

Comment (chrisnojima) [reply shared/reducers/chat2.js]: Is that the same key?

Comment (mmaxim) [reply shared/reducers/chat2.js]: The first one is in `metaMapReducer` the second is in `rootReducer`

**Comment by mmaxim** [comment]
Ok @joshblum this is ready for another look with the new method I described.

File: go/chat/sender.go [hunk 3]
```diff
@@ -437,6 +438,44 @@ func (s *BlockingSender) checkTopicNameAndGetState(ctx context.Context, msg chat
 	return topicNameState, nil
 }
 
+func (s *BlockingSender) resolveOutboxIDEdit(ctx context.Context, uid gregor1.UID,
+	convID chat1.ConversationID, msg *chat1.MessagePlaintext) error {
+	if msg.SupersedesOutboxID == nil {
+		return nil
+	}
+	s.Debug(ctx, "resolveOutboxIDEdit: resolving edit: outboxID: %s", msg.SupersedesOutboxID)
+	typ, err := msg.MessageBody.MessageType()
```
Comment (joshblum) [inline go/chat/sender.go]: small nit: can you add a `MessageBody.IsEdit()` method in extras? just cleaner to read, maybe avdl should just generate this..

File: go/chat/storage/outbox.go [hunk 4]
```diff
@@ -351,27 +351,43 @@ func (o *Outbox) RetryMessage(ctx context.Context, obid chat1.OutboxID,
 	return res, nil
 }
 
-func (o *Outbox) UpdateMessage(ctx context.Context, replaceobr chat1.OutboxRecord) error {
```
Comment (joshblum) [inline go/chat/storage/outbox.go]: revert these changes? seem unused

Comment (mmaxim) [reply go/chat/storage/outbox.go]: I kind of like them, I reverted the changes to `MessageDeliverer` though.

File: go/chat/sender.go (hunk 3 above)
Comment (mmaxim) [reply go/chat/sender.go]: Sure

Comment (mmaxim) [reply go/chat/sender.go]: Actually I kind of like having two different kinds of error here...

Comment (joshblum) [reply go/chat/sender.go]: ok we can leave as is
